### PR TITLE
fix(v2): fix too strict markdown frontmatter validation

### LINKED ISSUE #4642: 2.0.0-alpha.73 can't handle tags which could be parsed as numbers
## 🐛 Bug Report

As of 2.0.0-alpha.73, docusaurus will choke on tags that can be parsed as numbers.  

Consider the following:

```
---
title: "Azure DevOps Node API: The missing episodes"
author: John Reilly
author_url: https://github.com/johnnyreilly
author_image_url: https://blog.johnnyreilly.com/img/profile.jpg
tags: [azure devops api, 203, node.js]
hide_table_of_contents: false
---
```

The `203` tag above is bad news and will lead to an error like this:

```
[en] Creating an optimized production build...
error building locale=en
ValidationError: "tags[1]" must be one of [string, object]
```

See: https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/35/checks?check_run_id=2377642622

You can workaround it by changing `203` to be `'203'`, but ideally this wouldn't be necessary.

### Have you read the [Contributing Guidelines on issues]

Yes

## To Reproduce

Follow the example above and then `yarn build`

## Expected behavior

Compilation works.

## Actual Behavior

An error of the nature:

```
[en] Creating an optimized production build...
error building locale=en
ValidationError: "tags[1]" must be one of [string, object]
    at Object.exports.process (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/errors.js:184:16)
    at Object.internals.entry (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/validator.js:150:26)
    at Object.exports.entry (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/validator.js:27:30)
    at internals.Base.validate (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/base.js:548:26)
    at Object.internals.assert (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/index.js:225:27)
    at Object.attempt (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/joi/lib/index.js:107:26)
    at Object.assertBlogPostFrontMatter (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/plugin-content-blog/lib/blogFrontMatter.js:24:28)
    at /home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/plugin-content-blog/lib/blogUtils.js:90:27
    at async Promise.all (index 194)
    at async Object.generateBlogPosts (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/plugin-content-blog/lib/blogUtils.js:84:5)
    at async Object.loadContent (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/plugin-content-blog/lib/index.js:59:25)
    at async /home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/server/plugins/index.js:53:46
    at async Promise.all (index 0)
    at async Object.loadPlugins (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/server/plugins/index.js:52:34)
    at async Object.load (/home/runner/work/blog.johnnyreilly.com/blog.johnnyreilly.com/blog-website/node_modules/@docusaurus/core/lib/server/index.js:94:82)
```

## Your Environment

Locally in Linux and in GitHub Actions

- Public source code: https://github.com/johnnyreilly/blog.johnnyreilly.com
- Public site url: https://blog.johnnyreilly.com/
- Docusaurus version used: 2.0.0-alpha.73
- Environment name and version: Node.js 14
- Operating system and version: Desktop

## Reproducible Demo

See linked PR: https://github.com/johnnyreilly/blog.johnnyreilly.com/pull/35

I've opted to go with the workaround for now; but I thought I should share the experience.  Certainly the error message isn't the most helpful in the world.

### PR DESCRIPTION
**Pull request by johnnyreilly**
## Motivation

**PR completed by @slorber**:

- Add better markdown validation infra
- Convert numbers to strings automatically (#4642)
- Relax too strict string validation options (allow empty title/description...)

----

**Original description**: 

To either get a meaningful error message when invalid tags are supplied, or allow them to be parsed successfully (numbers are not) - this relates to #4642

Hey @slorber,

I've started work on this and have some questions.  Whilst I can get meaningful error messages using the mechanism you suggest (see changes in `blogFrontMatter.ts`) the suggested: 

```ts
    Joi.attempt(frontMatter, BlogFrontMatterSchema, {convert: true});
```

does not seem to support parsing numbers to strings automatically as suggested.  Also, even if it did, it won't actually change the type of the tags in the `frontMatter` object.  It's just an assertion - no mutation takes place.  We could put something in the `markdownParser` here:

https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-utils/src/markdownParser.ts#L68-L79

that migrates numbers to strings....  Or alternatively, maybe it just throwing a meaningful error message is fine.  I think the meaningful error message is the most significant thing from my perspective.  

What do you think?

### Have you read the [Contributing Guidelines on pull requests]

Yes

## Test Plan

Automated tests!

## Related PRs

N/A

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "Docusaurus Dev Container",
-  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:0-10-buster",
+  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:12-buster",
   "settings": {
     "terminal.integrated.shell.linux": "/bin/bash"
   },
-  "extensions": ["dbaeumer.vscode-eslint"],
+  "extensions": ["dbaeumer.vscode-eslint", "orta.vscode-jest"],
```
Comment (slorber) [inline .devcontainer/devcontainer.json:7]: I don't remember what this is about, but I trust you for this to be a good change :D

Can we use Node 14 instead?

**Review by slorber — COMMENTED** [review]
Thanks @johnnyreilly , I see the problem

I'd like to do a small bugfix release soon, and have a few issues reported regarding this frontmatter validation.

As this is likely to create conflicts (like with https://github.com/facebook/docusaurus/pull/4655), I'll take your PR as a base and complete it so that I can release the fix asap!

I changed the assertion to actually return the value. I will try to see if I can convert to number from there

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (johnnyreilly) [reply .devcontainer/devcontainer.json:7]: upgraded to node 14!  

The extension is https://marketplace.visualstudio.com/items?itemName=Orta.vscode-jest

Makes it much easier to work with Jest in VS Code - you don't have to take this but I think it's very useful!

image

Comment (slorber) [reply .devcontainer/devcontainer.json:7]: thanks, let's add this :)

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "Docusaurus Dev Container",
-  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:0-10-buster",
+  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:14-buster",
```
Comment (slorber) [inline .devcontainer/devcontainer.json:3]:
```suggestion
  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:14-buster", // test
```

**Pull request merged**